### PR TITLE
Fix issues with multiple translation units

### PR DESCRIPTION
**Pull request by nicbarker**
This PR addresses the issue raised in [this issue](https://github.com/nicbarker/clay/issues/19). I generally write my C applications as single translation units, so this was never tested. I basically copied the approach and rationale used in [Sean Barrett's STB libraries](https://github.com/nothings/stb?tab=readme-ov-file#how-do-i-use-these-libraries).

**This PR introduces a very small breaking change ONLY for direct C & C++ usage. Other language bindings are not affected.**

You must now `#define CLAY_IMPLEMENTATION` in **one** file somewhere, **before** you include `clay.h`:

```C
#define CLAY_IMPLEMENTATION
#include "clay.h"
```

I took the opportunity to improve the internal API somewhat, so less of it would need to be exposed. As long as you aren't depending on the private API, the changes should be transparent.

### CODE REVIEW

File: clay.h [hunk 1]
```diff
@@ -104,39 +82,367 @@
...
+// Internal API functions required by macros
+Clay_LayoutConfig * Clay__StoreLayoutConfig(Clay_LayoutConfig config);
+Clay_RectangleElementConfig * Clay__StoreRectangleElementConfig(Clay_RectangleElementConfig config);
+Clay_TextElementConfig * Clay__StoreTextElementConfig(Clay_TextElementConfig config);
+Clay_ImageElementConfig * Clay__StoreImageElementConfig(Clay_ImageElementConfig config);
+Clay_FloatingElementConfig * Clay__StoreFloatingElementConfig(Clay_FloatingElementConfig config);
+Clay_CustomElementConfig * Clay__StoreCustomElementConfig(Clay_CustomElementConfig config);
+Clay_ScrollElementConfig * Clay__StoreScrollElementConfig(Clay_ScrollElementConfig config);
+Clay_BorderElementConfig * Clay__StoreBorderElementConfig(Clay_BorderElementConfig config);
+Clay_ElementId Clay_HashString(Clay_String toHash, uint32_t index);
```
Comment (mikejsavage) [inline clay.h:400]: This is another double underscore name

File: clay.h [hunk 2]
```diff
@@ -104,39 +82,367 @@
...
+Clay_RenderCommandArray Clay_EndLayout();
+bool Clay_PointerOver(Clay_ElementId id);
+Clay_ScrollContainerData Clay_GetScrollContainerData(Clay_ElementId id);
+void Clay_SetMeasureTextFunction(Clay_Dimensions (*measureTextFunction)(Clay_String *text, Clay_TextElementConfig *config));
+Clay_RenderCommand * Clay_RenderCommandArray_Get(Clay_RenderCommandArray* array, int32_t index);
+void Clay_SetDebugModeEnabled(bool enabled);
+
+void Clay_OpenContainerElement(Clay_ElementId id, Clay_LayoutConfig *layoutConfig);
+void Clay_OpenRectangleElement(Clay_ElementId id, Clay_LayoutConfig *layoutConfig, Clay_RectangleElementConfig *rectangleConfig);
+void Clay_OpenTextElement(Clay_ElementId id, Clay_String *text, Clay_TextElementConfig *textConfig);
```
Comment (mikejsavage) [inline clay.h:381]: `text` should not be a pointer here

File: clay.h [hunk 3]
```diff
@@ -104,39 +82,367 @@
...
+#endif
+
+// -----------------------------------------
+// IMPLEMENTATION --------------------------
+// -----------------------------------------
+#ifdef CLAY_IMPLEMENTATION
+#undef CLAY_IMPLEMENTATION
+
+#ifdef CLAY_WASM
+#define CLAY_WASM_EXPORT(name) __attribute__((export_name(name)))
```
Comment (mikejsavage) [inline clay.h:415]: I guess this is part of the public API since the website uses it

File: clay.h [hunk 4]
```diff
@@ -360,67 +608,6 @@ Clay_RenderCommandType Clay__LayoutElementTypeToRenderCommandType[] = {
...
-
-typedef struct {
-    Clay_Sizing sizing;
-    Clay_Padding padding;
-    uint16_t childGap;
-    Clay_LayoutDirection layoutDirection;
-    Clay_ChildAlignment childAlignment;
-} Clay_LayoutConfig;
-
 Clay_LayoutConfig CLAY_LAYOUT_DEFAULT = (Clay_LayoutConfig){};
```
Comment (mikejsavage) [inline clay.h:611]: Website uses this too

**Review by mikejsavage — COMMENTED** [review]
This still has some issues and can't compile the website, I've added review comments instead of PRing since they're small and I don't know exactly how you want to fix them

Can't add a comment since it's not in the diff, but the website tries to set `Clay__debugViewHighlightColor` too

Also seeing a lot of internal stuff that could be marked `static` at this point

File: clay.h [hunk 5]
```diff
@@ -104,39 +82,367 @@
...
+void Clay_SetMeasureTextFunction(Clay_Dimensions (*measureTextFunction)(Clay_String *text, Clay_TextElementConfig *config));
+Clay_RenderCommand * Clay_RenderCommandArray_Get(Clay_RenderCommandArray* array, int32_t index);
+void Clay_SetDebugModeEnabled(bool enabled);
+
+void Clay_OpenContainerElement(Clay_ElementId id, Clay_LayoutConfig *layoutConfig);
+void Clay_OpenRectangleElement(Clay_ElementId id, Clay_LayoutConfig *layoutConfig, Clay_RectangleElementConfig *rectangleConfig);
+void Clay_OpenTextElement(Clay_ElementId id, Clay_String *text, Clay_TextElementConfig *textConfig);
+void Clay_OpenImageElement(Clay_ElementId id, Clay_LayoutConfig *layoutConfig, Clay_ImageElementConfig *imageConfig);
+void Clay_OpenScrollElement(Clay_ElementId id, Clay_LayoutConfig *layoutConfig, Clay_ScrollElementConfig *imageConfig);
+void Clay_OpenFloatingElement(Clay_ElementId id, Clay_LayoutConfig *layoutConfig, Clay_FloatingElementConfig *imageConfig);
```
Comment (mikejsavage) [inline clay.h:384]: These all have `imageConfig` as their last arg

File: clay.h (hunk 4 above)
Comment (nicbarker) [reply clay.h:611]: Fixed this here: https://github.com/nicbarker/clay/commit/5bb725d0057155200db13890a86922d393305167

File: clay.h (hunk 3 above)
Comment (nicbarker) [reply clay.h:415]: Moved these up to the forward declarations 👍

File: clay.h (hunk 2 above)
Comment (nicbarker) [reply clay.h:381]: Fixed 👍

File: clay.h (hunk 1 above)
Comment (nicbarker) [reply clay.h:400]: Fixed 👍

File: clay.h [hunk 6]
```diff
@@ -104,39 +82,367 @@
...
+void Clay_SetLayoutDimensions(Clay_Dimensions dimensions);
+void Clay_BeginLayout();
+Clay_RenderCommandArray Clay_EndLayout();
+bool Clay_PointerOver(Clay_ElementId id);
+Clay_ScrollContainerData Clay_GetScrollContainerData(Clay_ElementId id);
+void Clay_SetMeasureTextFunction(Clay_Dimensions (*measureTextFunction)(Clay_String *text, Clay_TextElementConfig *config));
+Clay_RenderCommand * Clay_RenderCommandArray_Get(Clay_RenderCommandArray* array, int32_t index);
+void Clay_SetDebugModeEnabled(bool enabled);
+
+void Clay_OpenContainerElement(Clay_ElementId id, Clay_LayoutConfig *layoutConfig);
```
[the remark this answers is not in the document]
Comment (nicbarker) [reply clay.h:379]: Fixed all of these 👍

File: clay.h (hunk 5 above)
Comment (nicbarker) [reply clay.h:384]: Fixed in https://github.com/nicbarker/clay/commit/6430bbbdc3e45fb7297acc530b1911bc7e49fc74

**Pull request merged**